File: android/src/main/java/com/imranmentese/reactnativenetwatch/model/NativeRequest.java
package com.imranmentese.reactnativenetwatch.model;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

import okhttp3.Response;

import static com.imranmentese.reactnativenetwatch.RNNetwatchModule.HTTP_COMPLETED;

/**
 * Created by MENTESE Imran on 16/02/2021.
 */

public class NativeRequest {
    int _id;
    String url;
    String method;
    int status;
    int readyState;
    long startTime;
    long endTime;
    Object requestHeaders;
    Object responseHeaders;
    String responseContentType;
    Object response;

    public NativeRequest(Response response, long startTime, long endTime) {
        if (response != null) {
            // From request
            this.method = response.request().method();
            this.url = response.request().url().url().toString();
            this.requestHeaders = response.request().headers();

            // From response
            this.status = response.code();

            Map<String, String> mapResponseHeaders = new HashMap<String, String>();
            for(String key: response.headers().names()){
                List<String> values = response.headers().values(key);
                if (values != null && values.size() > 0){
                    String value = values.get(0);
                    if (value.length() > 0 && (value.charAt(0) != '{'  || value.charAt(value.length() - 1) != '}')) {
                        mapResponseHeaders.put(key, value.replaceAll("\"", ""));
                    } else {
                        mapResponseHeaders.put(key, "\"" + value.replaceAll("\"", "") + "\"");
                    }
                }
            }
            this.responseHeaders = mapResponseHeaders;

            try {
                String responseBody = response.peekBody(Long.MAX_VALUE).string();
                if (responseBody != null) {
                    this.response = responseBody;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        this._id = new Random().nextInt(10000);
        this.startTime = startTime;
        this.endTime = endTime;
        this.readyState = HTTP_COMPLETED;
    }
}

File: ios/Model/NativeRequest.m
//
//  NativeRequest.m
//  RNNetwatch
//
//  Created by Imran Mentese on 09/03/2021.
//  Copyright © 2021 Imran Mentese. All rights reserved.
//

#import "NativeRequest.h"

@implementation NativeRequest

- (NSString *)formateRequestHeaderFieldKey:(NSString *)key object:(id)obj
{
    return [NSString stringWithFormat:@"%@:%@\n",key?:@"",obj?:@""];
}

- (NSDictionary*) toJSON {
    return @{
        @"_id": self._id ? [NSString stringWithFormat:@"%i", self._id] : @"",
        @"url": self.url ?: @"",
        @"method": self.method ?: @"",
        @"status": self.status ? [NSString stringWithFormat:@"%i", self.status] : @"",
        @"readyState": self.readyState ? [NSString stringWithFormat:@"%i", self.readyState] : @"",
        @"startTime": self.startTime ? [NSString stringWithFormat:@"%li", self.startTime] : @"",
        @"endTime": self.endTime? [NSString stringWithFormat:@"%li", self.endTime] : @"",
        @"requestHeaders": self.requestHeaders ?: @"",
        @"responseHeaders": self.responseHeaders ?: @"",
        @"responseContentType": self.responseContentType ?: @"",
        @"response": self.response ?: @""
    };
}

@end
